### PR TITLE
Execute terraform destroy

### PR DESCRIPTION
**Pull request by giorgiosironi**
When running `./bldr destroy:api-gateway--test`, I want not just the CloudFormation template but also the Fastly CDN to be destroyed as well.

### CODE REVIEW

**Comment by giorgiosironi** [comment]
Build hangs non-deterministically.

File: src/buildercore/bootstrap.py [hunk 1]
```diff
@@ -489,7 +450,7 @@ def update_stack(stackname, service_list=None, **kwargs):
         ('ec2', update_ec2_stack),
         ('s3', update_s3_stack),
         ('sqs', update_sqs_stack),
-        ('terraform', update_terraform_stack)
+        ('terraform', lambda stackname, context: terraform.update(stackname, context))
```
Comment (lsh-0) [inline src/buildercore/bootstrap.py]: this is redundant and can be rewritten like: `('terraform', terraform.update),`

File: src/buildercore/context_handler.py [hunk 2]
```diff
@@ -53,5 +53,18 @@ def download_from_s3(stackname, refresh=False):
...
     return True
 
+def only_if(servicename):
+    """Decorator that only executes an update function if the context contains a particular servicename that would need it"""
+    def decorate_with_only_if(fn):
+        def decorated_with_only_if(stackname, context, **kwargs):
+            # only update service if stack is using given service
+            LOG.info("Try to update '%s'", servicename)
+            if context.get(servicename):
+                # TODO: context is not always necessary in fn implementations. Can we avoid passing it when not needed?
```
Comment (lsh-0) [inline src/buildercore/context_handler.py]: it doesn't hurt either. avoiding passing the context would make the code unnecessarily more complicated.
https://docs.python.org/3/library/inspect.html#inspect.getargspec

**Comment by lsh-0** [comment]
I see you replacing `updates` with `only_if`, but I don't see the decorations being changed though (still `@updates(...)`). I don't know how the code is even running without taking a much closer look. I'll let you fix your other problem first

File: src/buildercore/bootstrap.py (hunk 1 above)
Comment (giorgiosironi) [reply src/buildercore/bootstrap.py]: nice that this isn't JavaScript :muscle:

**Comment by giorgiosironi** [comment]
> replacing updates with only_if, but I don't see the decorations being changed though

Should be covered by `from .context_handler import only_if as updates`

**Comment by giorgiosironi** [comment]
Reproduced the hanging at
```
2018-04-11 12:58:51,977 - INFO - DaemonlessPoolWorker-1 - buildercore.cfngen - validating crm, fresh
2018-04-11 12:58:52,563 - INFO - DaemonlessPoolWorker-1 - buildercore.cfngen - validating crm, 1604
```

**Comment by giorgiosironi** [comment]
The process tree looks like
```
14677 ?        S      0:00              \_ /bin/bash ./project_tests.sh
15346 ?        S      0:00                  \_ /usr/bin/python /usr/local/bin/tox
16773 ?        S      0:00                      \_ /bin/bash /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.run-tests.sh tests integration_tests
16958 ?        Sl     0:00                          \_ /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/python3.5 /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/green -vv --quiet-coverage integration_tests
16977 ?        R      4:18                              \_ /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/python3.5 /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/green -vv --quiet-coverage integration_tests
16978 ?        S      0:05                              \_ /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/python3.5 /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/green -vv --quiet-coverage integration_tests
16982 ?        Sl     0:00                              \_ /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/python3.5 /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/green -vv --quiet-coverage integration_tests
```

**Comment by giorgiosironi** [comment]
The time is increasing, so the Python code should be doing something, perhaps an infinite loop:
```
16977 ?        R      5:17      |                       \_ /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/python3.5 /ext/jenkins/workspace/er_PR-272-PRM3GT4Y3LMWBIJYXYSE/.tox/py35/bin/green -vv --quiet-coverage integration_tests
```
(it's at 100% CPU)

**Comment by giorgiosironi** [comment]
No idea why there are multiple processes involved, I would have expected everything to run into one by default as these tests are not designed for concurrency. Switching back to #273 to see if that helps.

**Comment by giorgiosironi** [comment]
Also verified manually that `destroy` shouldn't be requiring any user input.

**Comment by giorgiosironi** [comment]
@seanwiseman made this green, ready for review too now.

File: src/buildercore/terraform.py [hunk 3]
```diff
@@ -35,3 +40,33 @@ def render(context):
...
+
+@only_if('fastly')
+def update(stackname, context):
+    ensure('FASTLY_API_KEY' in os.environ, "a FASTLY_API_KEY environment variable is required to provision Fastly resources", ConfigurationError)
+    t = init(stackname)
+    t.apply(input=False, capture_output=False, raise_on_error=True)
+
+@only_if('fastly')
+def destroy(stackname, context):
+    t = init(stackname)
```
Comment (seanwiseman) [inline src/buildercore/terraform.py]: By default single char variable names should be flagged as `bad-names` by `pylint`. Can't see a `good-names` override in the `.pylintrc` or `.lint.sh`, it must be disabling the the rule somewhere 🤔 .

Would advise adhering to single chars being bad names especially if there are no type hints present.

File: src/tests/base.py [hunk 4]
```diff
@@ -41,7 +42,7 @@ def switch_out_test_settings(self):
     def generate_environment_name(self):
         """to avoid multiple people clashing while running their builds
            and new builds clashing with older ones"""
-        return check_output('whoami').rstrip().decode() + datetime.utcnow().strftime("%Y%m%d%H%M%S")
+        return "%s-%s-%s" % (check_output('whoami').rstrip().decode(), datetime.utcnow().strftime("%Y%m%d%H%M%S"), randint(1, 1000000))
```
[the remark this answers is not in the document]
Comment (giorgiosironi) [reply src/tests/base.py]: Nice one, it should be easier to join than to template here, as the only goal is to get a unique identifier

File: src/buildercore/terraform.py (hunk 3 above)
Comment (giorgiosironi) [reply src/buildercore/terraform.py]: will rename to `terraform` which doesn't shadow anything in this module

Comment (giorgiosironi) [reply src/buildercore/terraform.py]: To get it detected, I had to add `invalid-name` https://github.com/elifesciences/builder/blob/master/.lint.sh#L29
```
************* Module buildercore.terraform
C: 46, 4: Variable name "t" doesn't conform to snake_case naming style (invalid-name)
```
That however finds too many occurrences in other files, so should be changed separately.

**Pull request merged**